File: src/px4_vel_controller.cpp
/**
 * @file px4_vel_controller.cpp
 * @brief Node for converting velocity command of move_base to mavros
 * @ author Takaki Ueno
 */

// c++ library
#include <string>
// ROS
#include <ros/ros.h>
// TF
#include <tf/transform_listener.h>
// geometry_msgs
#include <geometry_msgs/Twist.h>
#include <geometry_msgs/TwistStamped.h>

//! mavros velocity publisher
ros::Publisher cmd_pub;

//! origin frame name
std::string origin_frame;
//! base frame name of vehicle
std::string vehicle_frame;

//! origin coordinate of origin_frame
tf::Vector3 global_origin(0, 0, 0);
//! pointer to tf listener
tf::TransformListener *listener_ptr;

/**
 * @brief Callback function for velocity command from move_base
 * @param msg Incoming msg
 */
void vel_cmd_cb(const geometry_msgs::Twist::ConstPtr &msg)
{
    tf::Vector3 vel_unstamped(msg->linear.x, msg->linear.y, 0);

    tf::StampedTransform vehicle_tf;
    try
    {
        // get tf from origin to base and transform velocity by using the tf
        listener_ptr->lookupTransform(origin_frame, vehicle_frame, ros::Time(0), vehicle_tf);
        vehicle_tf.setOrigin(global_origin);
        vel_unstamped = vehicle_tf * vel_unstamped;
    }
    catch (tf::TransformException& ex)
    {
        // print error and exit function without publishing message
        ROS_ERROR("%s", ex.what());
        return;
    }

    geometry_msgs::TwistStamped vel_cmd;

    // setting header field
    vel_cmd.header.frame_id = origin_frame;
    vel_cmd.header.stamp = ros::Time::now();

    // setting twist field
    vel_cmd.twist.linear.x = vel_unstamped.getX();
    vel_cmd.twist.linear.y = vel_unstamped.getY();
    vel_cmd.twist.linear.z = vel_unstamped.getZ();
    vel_cmd.twist.angular = msg->angular;

    cmd_pub.publish(vel_cmd);
}

int main(int argc, char **argv)
{
    ros::init(argc, argv, "px4_vel_controller");
    ros::NodeHandle nh;
    ros::Rate rate(20);

    // Fixed frame name
    nh.param<std::string>("origin_frame", origin_frame, "map");
    // Vehicle base name
    nh.param<std::string>("vehicle_frame", vehicle_frame, "base_link");

    // move_base velocity topic
    std::string move_base_topic;
    nh.param<std::string>("move_base_topic", move_base_topic, "/cmd_vel");
    // mavros velocity topic
    std::string mavros_topic;
    nh.param<std::string>("mavros_topic", mavros_topic, "/mavros/setpoint_velocity/cmd_vel");

    cmd_pub = nh.advertise<geometry_msgs::TwistStamped>(mavros_topic, 10);
    // subscriber for velocity command from move base
    ros::Subscriber cmd_sub = nh.subscribe<geometry_msgs::Twist>(move_base_topic, 10, vel_cmd_cb);

    // listener should initialized after ros::init
    tf::TransformListener listener;

    // use pointer to listener
    // so that listener can be used in callback function
    // and keep listener alive until the node dies
    // Listener starts buffering transform after created
    // and every query will fail if listener becomes out of scope
    listener_ptr = &listener;

    ros::spin();

    return 0;
}

File: script/px4_vel_controller.py
#!/usr/bin/env python

import rospy
import tf
from geometry_msgs.msg import Twist, TwistStamped, Vector3Stamped

class PX4VelocityController:

    def __init__(self):
        self.origin_frame = rospy.get_param("origin_frame",
                                            "map")
        self.vehicle_frame = rospy.get_param("vehicle_frame",
                                             "base_link")
        self.move_base_topic = rospy.get_param("move_base_topic",
                                               "/cmd_vel")
        self.mavros_topic = rospy.get_param("mavros_topic",
                                            "/mavros/setpoint_velocity/cmd_vel")

        self.transformer = tf.TransformListener()
        self.cmd_pub = rospy.Publisher(self.mavros_topic,
                                       TwistStamped,
                                       queue_size=10)
        self.cmd_sub = rospy.Subscriber(self.move_base_topic,
                                        Twist, self.vel_cmd_cb)

    def vel_cmd_cb(self, msg):
        linear_vel = Vector3Stamped()
        linear_vel.header.frame_id = self.vehicle_frame
        linear_vel.header.stamp = rospy.Time().now()

        linear_vel.vector.x = msg.linear.x
        linear_vel.vector.y = msg.linear.y
        linear_vel.vector.z = 0

        try:
            self.transformer.waitForTransform(self.vehicle_frame, self.origin_frame, rospy.Time().now(), rospy.Duration(0.05))
            linear_vel = self.transformer.transformVector3(self.origin_frame, linear_vel)
        except tf.Exception as e:
            rospy.logerr(e)
            return

        vel_cmd = TwistStamped()
        vel_cmd.header.stamp = rospy.Time().now()
        vel_cmd.header.frame_id = self.origin_frame

        vel_cmd.twist.linear.x = linear_vel.vector.x
        vel_cmd.twist.linear.y = linear_vel.vector.y
        vel_cmd.twist.linear.z = linear_vel.vector.z
        vel_cmd.twist.angular.z = msg.angular.z

        self.cmd_pub.publish(vel_cmd)


if __name__ == "__main__":
    rospy.init_node("px4_vel_controller")
    rate = rospy.Rate(20)
    PX4VelocityController()
    rospy.spin()